File: weatherapp/src/services/weatherService.ts
import type {ForecastData, WeatherData} from "../interfaces/weatherInterfaces.ts";

const API_KEY = import.meta.env.VITE_API_KEY;
const BASE_URL = import.meta.env.VITE_BASE_URL;

export async function getWeather(city: string): Promise<WeatherData> {
    const response = await fetch(BASE_URL + `/weather?q=${city}&appId=${API_KEY}&units=metric`);
    if (response.status == 404) throw new Error("City not found!");
    return await response.json() as Promise<WeatherData>;
}

export async function getForecast(city: string): Promise<ForecastData> {
    const response = await fetch(BASE_URL + `/forecast?q=${city}&appId=${API_KEY}&units=metric`);
    if (response.status == 404) throw new Error("City not found!");
    return await response.json() as Promise<ForecastData>;
}

File: 03-weather-cli/WeatherCLI/WeatherClasses/WeatherService.cs
using System.Net;
using System.Text.Json;

namespace WeatherCLI.WeatherClasses;

public class WeatherService
{

    private static readonly HttpClient _client = new HttpClient();
    private static readonly string _Base_URL = "https://api.openweathermap.org/data/2.5/";

    private static readonly JsonSerializerOptions Options = new JsonSerializerOptions
        { PropertyNameCaseInsensitive = true };

    public static async Task<WeatherResponse?> GetCurrentWeather(string city, string apiKey)
    {
        try
        {
            var response = await _client.GetAsync($"{_Base_URL}weather?q={city}&appId={apiKey}&units=metric");

            if (response.StatusCode == HttpStatusCode.NotFound)
            {
                Console.WriteLine("City not found");
                return null;
            }

            string json = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<WeatherResponse>(json, Options);
        }
        catch (HttpRequestException)
        {
            Console.WriteLine("No internet connection");
            return null;
        }
    }

    public static async Task<ForecastResponse?> GetForecast(string city, string apiKey)
    {
        try
        {
            var response = await _client.GetAsync($"{_Base_URL}forecast?q={city}&appId={apiKey}&units=metric");

            if (response.StatusCode == HttpStatusCode.NotFound)
            {
                Console.WriteLine("City not found");
                return null;
            }

            string json = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<ForecastResponse>(json, Options);
        }
        catch (HttpRequestException)
        {
            Console.WriteLine("No internet connection");
            return null;
        }
    }

}